File: backend/routes/stories.py
"""
Stories API routes
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from services.llm import LLMService

router = APIRouter()


class GenerateStoryRequest(BaseModel):
    topic: str
    level: str
    length: str = "Medium"
    theme: str = ""
    model: Optional[str] = None
    target_language: str = "German"


class SimplifyStoryRequest(BaseModel):
    text: str
    level: str
    model: Optional[str] = None
    target_language: str = "German"


class GenerateQuestionsRequest(BaseModel):
    text: str
    count: int = 3
    model: Optional[str] = None
    target_language: str = "German"


@router.post("/generate")
async def generate_story(request: GenerateStoryRequest):
    """Generate a language learning story"""
    try:
        result = await LLMService.generate_story(
            topic=request.topic,
            level=request.level,
            length=request.length,
            theme=request.theme,
            model=request.model,
            target_language=request.target_language
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/simplify")
async def simplify_story(request: SimplifyStoryRequest):
    """Simplify a story to a target level"""
    try:
        result = await LLMService.simplify_text(
            text=request.text,
            level=request.level,
            model=request.model,
            target_language=request.target_language
        )
        return {"simplified_text": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/questions")
async def generate_questions(request: GenerateQuestionsRequest):
    """Generate comprehension questions for a story"""
    try:
        result = await LLMService.generate_comprehension_questions(
            text=request.text,
            count=request.count,
            model=request.model,
            target_language=request.target_language
        )
        return {"questions": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: src/services/api/stories.js
/**
 * Stories API Service
 * Handles story generation, simplification, and question generation
 */

import { post } from './client';

/**
 * Generate a story using AI
 */
export async function generateStory({ topic, level, length = 'Medium', theme = '', model = null, targetLanguage = 'German' }) {
    try {
        const response = await post('/stories/generate', {
            topic,
            level,
            length,
            theme,
            model,
            target_language: targetLanguage,
        });
        return response;
    } catch (error) {
        console.error('Story generation failed:', error);
        throw error;
    }
}

/**
 * Simplify a story to a target level
 */
export async function simplifyStory({ text, level, model = null, targetLanguage = 'German' }) {
    try {
        const response = await post('/stories/simplify', {
            text,
            level,
            model,
            target_language: targetLanguage,
        });
        return response.simplified_text;
    } catch (error) {
        console.error('Story simplification failed:', error);
        throw error;
    }
}

/**
 * Generate comprehension questions for a story
 */
export async function generateQuestions({ text, count = 3, model = null, targetLanguage = 'German' }) {
    try {
        const response = await post('/stories/questions', {
            text,
            count,
            model,
            target_language: targetLanguage,
        });
        return response.questions;
    } catch (error) {
        console.error('Question generation failed:', error);
        throw error;
    }
}

export default {
    generateStory,
    simplifyStory,
    generateQuestions,
};
